File: backend/ShantiSangha.Friends/Models/ConnectionAttachment.cs
namespace ShantiSangha.Friends.Models;

/// Owner-private keepsake attached to a Connection — a photo, video,
/// PDF, or any other file the owner wants to keep alongside that
/// person. Visible only to the Connection's owner; the linked user
/// (if any) never sees it. `Kind` is auto-derived from `ContentType`
/// so the iOS UI can split MEDIA from FILES without trusting client
/// classification.
///
/// Storage lives in the shared friends-media S3 bucket under a
/// `connections/{ownerId}/{connectionId}/...` prefix so removing a
/// Connection makes orphan cleanup straightforward.
public class ConnectionAttachment
{
    public Guid Id { get; set; }
    public Guid ConnectionId { get; set; }
    public Guid OwnerUserId { get; set; }
    public string ObjectKey { get; set; } = string.Empty;
    public string ContentType { get; set; } = string.Empty;
    public long ByteSize { get; set; }
    public string FileName { get; set; } = string.Empty;
    public ConnectionAttachmentKind Kind { get; set; } = ConnectionAttachmentKind.File;
    public string? Caption { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
}

File: ios/ShantiSangha/Models/ConnectionAttachment.swift
import Foundation

/// Mirror of the backend `ConnectionAttachmentResponse`. Owner-private
/// keepsake (photo, video, file) attached to a single Connection.
/// `kind` is server-assigned from the content type so the iOS UI can
/// split MEDIA from FILES without trusting itself. `downloadUrl` is a
/// short-lived presigned GET — refetch the list to refresh.
struct ConnectionAttachment: Codable, Identifiable, Equatable, Hashable {
    let id: UUID
    let connectionId: UUID
    let objectKey: String
    let contentType: String
    let byteSize: Int64
    let fileName: String
    let kind: ConnectionAttachmentKind
    let caption: String?
    let downloadUrl: String
    let createdAt: String
    let updatedAt: String

    /// Stable identity for SwiftUI `task(id:)` reloads — combines the
    /// stable id with the rotating download URL so the avatar-style
    /// "refetch when URL changes" pattern keeps working.
    var taskKey: String { "\(id.uuidString):\(downloadUrl)" }
}

enum ConnectionAttachmentKind: String, Codable {
    case media = "Media"
    case file = "File"
}

struct CreateConnectionAttachmentUploadRequest: Encodable {
    let contentType: String
    var fileName: String? = nil
}

struct CreateConnectionAttachmentUploadResponse: Decodable {
    let objectKey: String
    let uploadUrl: String
    let uploadUrlExpiresAt: String
}

struct CommitConnectionAttachmentRequest: Encodable {
    let objectKey: String
    let contentType: String
    let fileName: String
    var caption: String? = nil
}

struct UpdateConnectionAttachmentRequest: Encodable {
    /// Empty string clears the caption; nil leaves it unchanged.
    var caption: String? = nil
}
